server/Game.js

```diff
@@ -171,17 +158,33 @@ export default class {
     }
   }
 
+  getShuffledCandidates = (originalArray, cycles = MAX_CANDIDATES) => {
+    const shuffledArray = [...originalArray];
+    const cycleLimit = (shuffledArray.length - 1) - cycles;
+    
+    for (let i = shuffledArray.length - 1; i > 0 || i > cycleLimit; i--) {
+      const randomIndex = Math.floor(Math.random() * (i + 1));
+      const refWordSwaped = shuffledArray[i];
+      shuffledArray[i] = shuffledArray[randomIndex];
+      shuffledArray[randomIndex] = refWordSwaped;
+    }
+  
+    return shuffledArray.slice(-cycles);
+  }
+
   sendWordCandidates() {
     const player = this.nextPlayer();
+    const shouldClearWordsPlayed =
+    this.words.size > words.length / 2 ||
+    words.length - this.words.size < MAX_CANDIDATES;
 
     if (!player) return;
-
     this.currentPlayer = player.id;
     
-    if (this.words.size > words.length / 2) this.words.clear();
+    if (shouldClearWordsPlayed) this.words.clear();
 
     const availableWords = words.filter(word => !this.words.has(word));
-    this.currentCandidates = shuffleArray(availableWords).slice(0, MAX_CANDIDATES);
+    this.currentCandidates = this.getShuffledCandidates(availableWords);
```

Comment:
Any reason this is now a member function of the class instead of a free function? If we keep it like this, we could even inline the `availableWords` to avoid an extra copy of the words:

```js
getShuffledCandidates() {
  const shuffledArray = [words.filter(word => !this.words.has(word));
  ...
```

Reply:
I made this change because now the function is connected with the logic of the class/cadidates. It is no longer a generic function that shuffles any array. I liked your suggestion. 💜